train.py

```diff
@@ -0,0 +1,491 @@
+            model = model.to(device="cpu")
+
+        with h5py.File(initcache, mode='r') as h5:
+            clsts = h5.get("centroids")[...]
+            traindescs = h5.get("descriptors")[...]
+            model.pool.init_params(clsts, traindescs)
+            del clsts, traindescs
+
+        print('debugpoint')
+
+    isParallel = False
+    if int(config['global_params']['nGPU']) > 1 and torch.cuda.device_count() > 1:
+        model.encoder = nn.DataParallel(model.encoder)
+        model.pool = nn.DataParallel(model.pool)
+        isParallel = True
+
+    if config['train']['optim'] == 'ADAM':
+        optimizer = optim.Adam(filter(lambda par: par.requires_grad,
+                                      model.parameters()), lr=float(config['train']['lr']))  # , betas=(0,0.9))
+    elif config['train']['optim'] == 'SGD':
+        optimizer = optim.SGD(filter(lambda par: par.requires_grad,
+                                     model.parameters()), lr=float(config['train']['lr']),
+                              momentum=float(config['train']['momentum']),
+                              weight_decay=float(config['train']['weightDecay']))
+
+        scheduler = optim.lr_scheduler.StepLR(optimizer, step_size=int(config['train']['lrstep']),
+                                              gamma=float(config['train']['lrgamma']))
+    else:
+        raise ValueError('Unknown optimizer: ' + config['train']['optim'])
+
+    criterion = nn.TripletMarginLoss(margin=float(config['train']['margin']) ** 0.5, p=2, reduction='sum').to(device)
+
+    model = model.to(device)
+
+    if opt.resume_path:
+        optimizer.load_state_dict(checkpoint['optimizer'])
+
+    print('===> Loading dataset(s)')
+    exlude_panos_training = not config['train'].getboolean('includepanos')
+    train_dataset = MSLS(opt.dataset_root_dir, mode='train', transform=input_transform(),
+                         bs=int(config['train']['cachebatchsize']), threads=opt.threads, margin=float(config['train']['margin']),
+                         exclude_panos=exlude_panos_training)
+
+    validation_dataset = MSLS(opt.dataset_root_dir, mode='val', transform=input_transform(),
+                              bs=int(config['train']['cachebatchsize']), threads=opt.threads,
+                              margin=float(config['train']['margin']), posDistThr=25)
+
+    print('===> Training query set:', len(train_dataset.qIdx))
+    print('===> Evaluating on val set, query count:', len(validation_dataset.qIdx))
+    print('===> Training model')
+    writer = SummaryWriter(
+        log_dir=join(opt.save_path, datetime.now().strftime('%b%d_%H-%M-%S') + '_' + opt.identifier))
+
+    # write checkpoints in logdir
+    logdir = writer.file_writer.get_logdir()
+    opt.save_file_path = join(logdir, 'checkpoints')
+    makedirs(opt.save_file_path)
+
+    not_improved = 0
+    best_score = 0
+    if opt.resume_path:
+        # noinspection PyUnboundLocalVariable
+        if 'not_improved' in checkpoint:
```

Comment:
I think this can be removed - `not_improved` should always be in the checkpoint now.

Reply:
Yep, I think so too